### PR TITLE
Grpc Support 

### PR DESCRIPTION
**Pull request by NeGnail**
### Motivation:
Support GRPC.

### Modification:
Provides the GRPC bootstrap(consumer/provider), server, client transport and other abstractions
Add proxy for GRPC base class.
The stub for GRPC is abstracted mainly at the Transport layer.
update sofa-tracer version to 2.1.2.
update guava version to 20.0.

### Result:
Trace back to the #57
Docking of the main modes of use. Set up the basic architecture for GRPC's more complete extension in sofarpc.
Next we can support tracer and client-side streaming calls.

### CODE REVIEW

**Comment by khotyn** [comment]
文档的 PR 也请同步提交哈。 @NeGnail

**Comment by NeGnail** [comment]
恩呐，文档也会提的。

**Comment by NeGnail** [comment]
GRPC需要jdk7以上版本。  这里CI跑jdk6所以挂了。

**Comment by leizhiyuan** [comment]
https://github.com/alipay/sofa-rpc/issues/384

**Comment by NeGnail** [comment]
ok。那这个版本开始就可以去掉6的ci了。

**Comment by leizhiyuan** [comment]
恩，等下周5.4.8 发布后，有了 jdk8的分支，这个再合并。

**Comment by NeGnail** [comment]
暂时别合。今天我优化一下。

**Comment by leizhiyuan** [comment]
> 暂时别合。今天我优化一下。

嗯，只是解决了一下冲突。和hytrix和nocos的代码pom有冲突。

File: core-impl/proxy/src/main/java/com/alipay/sofa/rpc/proxy/javassist/JavassistProxy.java [hunk 1]
```diff
@@ -66,6 +69,41 @@
      */
     private static final Map<Class, Class> PROXY_CLASS_MAP = new ConcurrentHashMap<Class, Class>();
 
+    public <T> T getProxyForClass(final Class<T> clazz, final Invoker proxyInvoker) {
```
Comment (ujjboy) [inline core-impl/proxy/src/main/java/com/alipay/sofa/rpc/proxy/javassist/JavassistProxy.java]: 这个方法是为了？

**Comment by zonghaishang** [comment]
@NeGnail 解决下冲突

File: extension-impl/remoting-grpc/src/main/java/com/alipay/sofa/rpc/server/grpc/GrpcServer.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+    @Override
+    public void stop() {
+        if (!started) {
+            return;
+        }
+        try {
+            // 关闭端口，不关闭线程池
+            if (LOGGER.isInfoEnabled()) {
+                LOGGER.info("Stop the http rest server at port {}", serverConfig.getPort());
```
Comment (zonghaishang) [inline extension-impl/remoting-grpc/src/main/java/com/alipay/sofa/rpc/server/grpc/GrpcServer.java]: should change rest to grpc ??

File: extension-impl/remoting-grpc/src/main/java/com/alipay/sofa/rpc/server/grpc/GrpcServer.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+            }
+            server.shutdown();
+        } catch (Exception e) {
+            LOGGER.error("Stop the http rest server at port " + serverConfig.getPort() + " error !", e);
```
Comment (zonghaishang) [inline extension-impl/remoting-grpc/src/main/java/com/alipay/sofa/rpc/server/grpc/GrpcServer.java]: should change rest to grpc ??

File: extension-impl/remoting-grpc/src/main/java/com/alipay/sofa/rpc/server/grpc/GrpcServer.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    @Override
+    public void start() {
+        if (started) {
+            return;
+        }
+        synchronized (this) {
+            try {
+                server.start();
+                if (LOGGER.isInfoEnabled()) {
+                    LOGGER.info("Start the http rest server at port {}", serverConfig.getPort());
```
Comment (zonghaishang) [inline extension-impl/remoting-grpc/src/main/java/com/alipay/sofa/rpc/server/grpc/GrpcServer.java]: should change rest to grpc ??

**Pull request closed without merging**